File: schemas/crm/convex/tasks.ts
// tasks: actionable items assigned to users, leads, or deals with priority and status tracking.
// See README.md for full design rationale.

import { defineTable } from "convex/server";
import { v } from "convex/values";

export const tasks = defineTable({
  title: v.string(),
  description: v.optional(v.string()),
  dueDate: v.optional(v.string()),
  priority: v.union(v.literal("low"), v.literal("medium"), v.literal("high")),
  status: v.union(
    v.literal("todo"),
    v.literal("in_progress"),
    v.literal("completed"),
    v.literal("cancelled")
  ),
  completedAt: v.optional(v.number()),
  contactId: v.optional(v.id("contacts")),
  companyId: v.optional(v.id("companies")),
  dealId: v.optional(v.id("deals")),
  leadId: v.optional(v.id("leads")),
  ownerId: v.id("users"),
  updatedAt: v.number(),
})
  .index("by_owner_id_and_status", ["ownerId", "status"])
  .index("by_due_date", ["dueDate"])
  .index("by_contact_id", ["contactId"])
  .index("by_deal_id", ["dealId"])
  .index("by_company_id", ["companyId"])
  .index("by_lead_id", ["leadId"]);

File: schemas/crm/spacetimedb/tasks.rs
// tasks: Action items and to-dos assigned to CRM users.
// See README.md for full design rationale.

use spacetimedb::Timestamp;

#[derive(SpacetimeType, Clone)]
pub enum TaskPriority {
    Low, // type: String
    Medium,
    High,
}

#[derive(SpacetimeType, Clone)]
pub enum TaskStatus {
    Todo, // type: String
    InProgress,
    Completed,
    Cancelled,
}

#[spacetimedb::table(name = tasks, public)]
pub struct Task {
    #[primary_key]
    pub id: String, // UUID
    pub title: String,
    pub description: Option<String>,
    #[index(btree)]
    pub due_date: Option<String>, // YYYY-MM-DD
    pub priority: TaskPriority,
    pub status: TaskStatus,
    pub completed_at: Option<Timestamp>,
    #[index(btree)]
    pub contact_id: Option<String>, // UUID — FK → contacts.id (set null)
    #[index(btree)]
    pub company_id: Option<String>, // UUID — FK → companies.id (set null)
    #[index(btree)]
    pub deal_id: Option<String>, // UUID — FK → deals.id (set null)
    #[index(btree)]
    pub lead_id: Option<String>, // UUID — FK → leads.id (set null)
    #[index(btree)]
    pub owner_id: String, // UUID — FK → users.id (cascade delete)
    pub created_at: Timestamp,
    pub updated_at: Timestamp,
}
